### DISCUSSION TITLE
i18n for Captive Portal

### DISCUSSION THREAD

**Discussion by tiefenauer** [gitbls/sdm#53] [Ideas]
I would like to contribute to this project by adding to the Captive Portal. I am a native German speaker and also speak French and a little Italian. I could at least provide translations in German in high quality.

From what I see all logic including the templates are inside [sdm-cportal](https://github.com/gitbls/sdm/blob/master/sdm-cportal). I would like to collect some ideas, especially from @gitbls before hacking away.

The current implementation is simple and also consistent with the rest of sdm where each functionality is encapsulated in a single file. However, if i18n becomes a topic, I don't think this approach will scale well.

Here's what I would like to see in an MVP:

- Include a language switcher for the languages supported
- replace the texts in [sdm-cportal](https://github.com/gitbls/sdm/blob/master/sdm-cportal) with placeholders using Python's string interpolation
- provide an i18n file with the texts as a key=value for each language supported in a separate subfolder
- dynamically load the translations for the language selected from the respective file and populate python variables with it (before assigning the variables containing the HTML

Pros:
- the HTML would serve as a template and stilll be contained inside a file
- Captive Portal is a really cool feature that will be available to more users who are not native English speakers

Cons:
- more files than just a single one
- i18n files should not contain markup as it will be hard to maintain a valid DOM like this

Questions:
- What subfolder should the i18n files go into?

What do you think?

**Comment by gitbls** [comment]
TBH I don't think that sdm-cportal is used much. It's a bit awkward to use with only one WiFi adapter, and I'm sure that there are a few issues lurking in the code.

I'd encourage you to work with it a bit to get comfortable with it (you'll need to do this anyhow if you're going to work on it).

That said, if you find that it's good enough to work on further, here's how I'd think about handling i18n:
- Add a `--i18n` switch to the sdm-cportal command line that specifies the language to use. Defaults to English
- Keep the translations in /usr/local/sdm/i18n. Filenames should indicate the language (e.g., messages.english, messages.french or whatever) so that the command parsing code can easily determine if the requested language is supported

**Reply by tiefenauer** [reply]
Alright, I'll decide whether investing time in it makes sense after I played around with it for a while. For me personally this is a very nice addition to sdm's core functionality but I might opt for only implementing what I need in my fork.

**Reply by tiefenauer** [reply]
I tried out the captive portal a bit and didn't get it to work. The AP is created and the form to enter the network parameters is displayed on my mobile phone. However, after entering the SSID and password, the server seems to be shut down and the crucial second step which should test the parameters cannot be triggered. Network connection can therefore not be established, leaving the raspberry in an offline state. 

Not sure if this is related to the issues lurking in the code you referred to our to the fact I'm only using the raspberry's built-in WiFi adapter.

I decided to use [WiFi-Connect](https://github.com/balena-os/wifi-connect) after all, which does not seem to be too bulky and offers some nice additions (list of available networks is automatically created, UI can be easily customized)
